sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/customtokenproxy/CustomTokenProxyHttpClient.java

```diff
@@ -0,0 +1,248 @@
+import com.azure.identity.implementation.util.IdentitySslUtil;
+
+import reactor.core.publisher.Mono;
+
+public class CustomTokenProxyHttpClient implements HttpClient {
+
+    private static final ClientLogger LOGGER = new ClientLogger(CustomTokenProxyHttpClient.class);
+
+    private final ProxyConfig proxyConfig;
+    private volatile SSLContext cachedSSLContext;
+    private volatile byte[] cachedFileContent;
+    private final URL proxyUrl;
+    private final String sniName;
+    private final byte[] caData;
+    private final String caFile;
+
+    public CustomTokenProxyHttpClient(ProxyConfig proxyConfig) {
+        this.proxyConfig = proxyConfig;
+        this.proxyUrl = proxyConfig.getTokenProxyUrl();
+        this.sniName = proxyConfig.getSniName();
+        this.caData = proxyConfig.getCaData();
+        this.caFile = proxyConfig.getCaFile();
+    }
+
+    @Override
+    public Mono<HttpResponse> send(HttpRequest request) {
+        return Mono.fromCallable(() -> sendSync(request, Context.NONE));
+    }
+
+    @Override
+    public HttpResponse sendSync(HttpRequest request, Context context) {
+        try {
+            HttpURLConnection connection = createConnection(request);
+            return new CustomTokenProxyHttpResponse(request, connection);
+        } catch (IOException e) {
+            throw LOGGER.logExceptionAsError(new RuntimeException("Failed to create connection to token proxy", e));
+        }
+    }
+
+    private HttpURLConnection createConnection(HttpRequest request) throws IOException {
+        URL updatedUrl = rewriteTokenRequestForProxy(request.getUrl());
+        HttpsURLConnection connection = (HttpsURLConnection) updatedUrl.openConnection();
+        // If SNI explicitly provided
+        try {
+            SSLContext sslContext = getSSLContext();
+            SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();
+            if (!CoreUtils.isNullOrEmpty(sniName)) {
+                sslSocketFactory = new IdentitySslUtil.SniSslSocketFactory(sslSocketFactory, sniName);
+            }
+            connection.setSSLSocketFactory(sslSocketFactory);
+            connection.setHostnameVerifier(sniAwareVerifier(sniName, proxyUrl));
+        } catch (Exception e) {
+            throw LOGGER.logExceptionAsError(new RuntimeException("Failed to set up SSL context for token proxy", e));
+        }
+
+        String method = request.getHttpMethod().toString();
+        connection.setRequestMethod(method);
+        connection.setInstanceFollowRedirects(false);
+        connection.setConnectTimeout(10_000);
+        connection.setReadTimeout(20_000);
+
+        boolean hasBody = request.getBodyAsBinaryData() != null && request.getBodyAsBinaryData().toBytes() != null;
+
+        if (hasBody && ("POST".equals(method) || "PUT".equals(method) || "PATCH".equals(method))) {
+            connection.setDoOutput(true);
+        } else {
+            connection.setDoOutput(false);
+        }
```

Comment:
do we need to enforce an opinion on whether the request should have a body?
```suggestion
        connection.setDoOutput(hasBody);
```